crates/sequencer/src/block_template/manager.rs

```diff
@@ -0,0 +1,119 @@
+use std::{
+    collections::{
+        hash_map::Entry::{Occupied, Vacant},
+        HashMap,
+    },
+    sync::Arc,
+};
+
+use strata_crypto::verify_schnorr_sig;
+use strata_db::traits::Database;
+use strata_eectl::engine::ExecEngineCtl;
+use strata_primitives::{
+    block_credential::CredRule,
+    l2::L2BlockId,
+    params::{Params, RollupParams},
+};
+use strata_state::{
+    block::L2BlockBundle,
+    header::{L2BlockHeader, L2Header},
+};
+use strata_status::StatusChannel;
+
+use crate::block_template::{BlockCompletionData, BlockTemplate, BlockTemplateFull, Error};
+
+#[derive(Debug)]
+pub struct BlockTemplateManager<D, E> {
+    pub(crate) params: Arc<Params>,
+    pub(crate) database: Arc<D>,
+    pub(crate) engine: Arc<E>,
+    pub(crate) status_channel: StatusChannel,
+    // TODO: add some form of expiry to purge orphaned items
+    pending_templates: HashMap<L2BlockId, BlockTemplateFull>,
+    pending_by_parent: HashMap<L2BlockId, L2BlockId>,
+}
+
+impl<D, E> BlockTemplateManager<D, E>
+where
+    D: Database,
+    E: ExecEngineCtl,
+{
+    pub fn new(
+        params: Arc<Params>,
+        database: Arc<D>,
+        engine: Arc<E>,
+        status_channel: StatusChannel,
+    ) -> Self {
+        Self {
+            params,
+            database,
+            engine,
+            status_channel,
+            pending_templates: HashMap::new(),
+            pending_by_parent: HashMap::new(),
+        }
+    }
+
+    pub(crate) fn insert_template(&mut self, block_id: L2BlockId, template: BlockTemplateFull) {
+        let parent = *template.header().parent();
+        self.pending_templates.insert(block_id, template);
+        self.pending_by_parent.insert(parent, block_id);
```

Comment:
Maybe this should print a warning when evicting a child?  Or do some other handling if we're doing something weird with conflicting templates so we at least can know if it's happening.